locations/jclouds/src/main/java/brooklyn/location/jclouds/JcloudsLocation.java

```diff
@@ -1305,118 +1326,148 @@ public UserCreation(LoginCredentials creds, List<Statement> statements) {
                 }
             }
             
-        } else if (user.equals(ROOT_USERNAME)) {
-            // Authorizes the public-key and sets password for the root user, so can login as expected
-            if (password != null) {
-                statements.add(new ReplaceShadowPasswordEntry(Sha512Crypt.function(), ROOT_USERNAME, password));
-                loginCreds = LoginCredentials.builder().user(user).password(password).build();
-            }
-            if (publicKeyData!=null) {
-                statements.add(new AuthorizeRSAPublicKeys("~"+ROOT_USERNAME+"/.ssh", ImmutableList.of(publicKeyData)));
-                if (privateKeyData != null) {
-                    loginCreds = LoginCredentials.builder().user(user).privateKey(privateKeyData).build();
+        } else {
+            String pubKey = credential.getPublicKeyData();
+            String privKey = credential.getPrivateKeyData();
+            
+            if (credential.get()==null) {
+                if (!loggedSshKeysHint && !config.containsKey(PRIVATE_KEY_FILE)) {
+                    loggedSshKeysHint = true;
+                    LOG.info("Default SSH keys not found or not usable; will create new keys for each machine. "
+                        + "Create ~/.ssh/id_rsa or "
+                        + "set "+PRIVATE_KEY_FILE.getName()+" / "+PRIVATE_KEY_PASSPHRASE.getName()+" / "+PASSWORD.getName()+" "
+                        + "as appropriate for this location if you wish to be able to log in without Brooklyn.");
                 }
+                KeyPair newKeyPair = SecureKeys.newKeyPair();
+                pubKey = SecureKeys.toPub(newKeyPair);
+                privKey = SecureKeys.toPem(newKeyPair);
+                LOG.debug("Brooklyn key being created for "+user+" at new machine "+this+" is:\n"+privKey);
             }
+            // ensure credential is not used any more, as we have extracted al useful info
+            credential = null;
             
-        } else {
             // Create the user
             // note AdminAccess requires _all_ fields set, due to http://code.google.com/p/jclouds/issues/detail?id=1095
             AdminAccess.Builder adminBuilder = AdminAccess.builder()
                     .adminUsername(user)
-                    .adminPassword(password)
-                    .grantSudoToAdminUser(groovyTruth(grantUserSudo))
-                    .resetLoginPassword(true)
-                    .loginPassword(password);
-
-            if (publicKeyData!=null) {
-                adminBuilder.authorizeAdminPublicKey(true).adminPublicKey(publicKeyData);
+                    .grantSudoToAdminUser(groovyTruth(grantUserSudo));
+                    // login user's password now set to something random
+                    // (could explicitly set with loginPassword)
+            
+            Boolean useKey = null;
+            if (passwordToSet!=null) {
+                adminBuilder.adminPassword(passwordToSet);
+                useKey = false;
```

Comment:
This looks wrong, so not sure how it works. Presumably we'll always go into this if branch because `passwordToSet` is never null. Therefore we'll always set `userKey = false`.